widget/singleruneentry.go

```diff
@@ -0,0 +1,124 @@
+package widget
+
+import (
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/widget"
+)
+
+// Lowercase represents all lowercase characters
+var Lowercase = []rune{'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'}
+
+// Uppercase represents all uppercase characters
+var Uppercase = []rune{'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'}
+
+// Digits represents all digits.
+var Digits = []rune{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'}
+
+// Special represents all special characters
+var Special = []rune{'`', '~', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '-', '_', '=', '+', '[', ']', '\\', ';', ':', '\'', '"', ',', '<', '.', '>', '?', '/'}
+
+// Letters includes all lower and upper case letters.
+var Letters = append(Lowercase, Uppercase...)
+
+// LettersAndDigits includes all letters and digits.
+var LettersAndDigits = append(Letters, Digits...)
+
+// AllRunes contains all possibule runes.
+var AllRunes = append(LettersAndDigits, Special...)
+
+// SingleRuneEntry is an entry that only allows a single rune
+// from an optional list of allowed runes. It exposes methods
+// to respond to navigation events while the entry is in focus.
+type SingleRuneEntry struct {
+	widget.Entry
+
+	// OnNavigate is called under the following circumstances:
+	//
+	//  KeyLeft/KeyRight
+	//    - Called with the corresponding key
+	//
+	//  Backspace
+	//    - Called with the key if the length of this entry is
+	//      already zero.
+	//
+	//  TextAdded
+	//    - Called with key set to a nil value
+	//      TODO: Probably have a separate OnChanged or fall up to
+	//      the widget.Entry implementation. Or declare new typed
+	//      constants.
+	OnNavigate func(key *fyne.KeyEvent)
```

Comment:
I don't think this method is needed - if developers have to filter on the key they would be just as well to extend and use KeyTyped like with a regular entry.

Reply:
I must have missed `KeyTyped` - I'll look into that because it sounds like it would simplify things.

Reply:
Sorry, I typeoed - I meant TypedKey, like you already use in the code

Reply:
I'm trying other ways of implementing this, and I'm wondering if it is better for it to not be an exported widget. The OnNavigate serves the focusing functionality in the `SegmentedEntry`. I'm worried removing it here will just lead to more complexity there. Though the widget _could_ see use for other forms where a user just wants to accept a single rune, it wasn't the original intention.

Thoughts?

Reply:
Though on second thought, the aforementioned user wanting an entry which accepts a single rune...they could use a SegmentedEntry with length set to 1.